### DISCUSSION TITLE
Auto Reply OOO Debugging

### DISCUSSION THREAD

**Discussion by jamesarbrown** [Mailu/Mailu#3159] [User support]
Have mailu 2.0.30 running under Kubernetes. Everything is fine apart from auto reply. Emails goto user, just no auto reply comes back.

Found this in the dovecot logs.
```Error: smtp-client: conn mailu-front.mailu200124.svc.cluster.local:25 (10.43.96.92:25) [1]: Connection timed out after 30 seconds```

followed by 
```
lmtp(xyz@example.com)<2734><vb5UMdk6xWWuCgAAmEiwvg>: Error: sieve: msgid=<36qkiualjv43dfdrsm2edugg.1707424471129@gmail.com>: vacation action: failed to send vacation response to SRS0=lNgX=JR=gmail.com=user@example.uk: <smtp(mailu-front.mailu200124.svc.cluster.local): RCPT TO failed: Connect timed out> (temporary error)
```
Could anyone advise how we can debug this further

**Comment by jamesarbrown** [comment]
To check the IP and Pod were actually there on SMTP I fired up a networktools pod and ran
```
networktools:/# telnet mailu-front.mailu200124.svc.cluster.local 25
Trying 10.43.96.92...
Connected to mailu-front.mailu200124.svc.cluster.local.
Escape character is '^]'.
```

So it appear the pod is there and port open. No idea why "smtp-client" can not reach it.

**Reply by jamesarbrown** [reply]
Still trying to figure this one. There is a problem with the SSL which maybe the root of it.

This is dovecot logs
```
Feb 19 20:52:10 lmtp(1415): Error: SSL context initialization failed, disabling SSL: Can't load SSL certificate (ssl_cert setting): The certificate is empty 
Feb 19 20:52:10 lmtp(1415): Info: Connect from 10.42.1.56 
Feb 19 20:52:40 lmtp(reply@example.com)<1415><8f8hD3q/02WHBQAAmEiwvg>: Error: smtp-client: conn mailu-front.mailu200124.svc.cluster.local:25 (10.43.96.92:25) [1]: Connection timed out after 30 seconds 
Feb 19 20:52:40 lmtp(reply@example.com)<1415><8f8hD3q/02WHBQAAmEiwvg>: Error: sieve: msgid=<5neijiclva2tcj3l2rivdm07.1708375927008@example.com>: vacation action: failed to send vacation response to james@example.com: <smtp(mailu-front.mailu200124.svc.cluster.local): RCPT TO failed: Connect timed out> (temporary error) 
Feb 19 20:52:40 lmtp(reply@example.com)<1415><8f8hD3q/02WHBQAAmEiwvg>: Info: sieve: msgid=<5neijiclva2tcj3l2rivdm07.1708375927008@example.com>: stored mail into mailbox 'INBOX' 
Feb 19 20:52:40 lmtp(1415): Info: Disconnect from 10.42.1.56: Logged out (state=READY)
```

So i ran from another pod to check the cert
```
openssl s_client -connect mailu-front.mailu200124.svc.cluster.local:993
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 343 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
This TLS version forbids renegotiation.
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```
As the logs say, an empty certificate.

But, if as imap.example.org, via a traefik TCP loadbalancer (but **NO** TLS endpoint on traefik), i get a proper cert.
```
openssl s_client -connect imap.example.org:993 |more
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = example.org
verify return:1
CONNECTED(00000003)

...
```

In helm the ingress is set to true and the certificate set (certificate is by cert-manager for the entire cluster and reflected into this namespace)
```
  tls: true
  existingSecret: "example-org-tls"
```

In the front pod the certificates are present in /certs (have checked the keys also are the right ones).
```
mailu-front-5656fc7dc9-5n58p:/etc/nginx# ls /certs
cert.pem  key.pem
```

These are refered to by /etc/nginx/tls.conf

So at a bit of a loss now, the certificate is present, but being served up as empty when you hit the pod directly via its URI and not the mail server URI. TBH the vacation would be better hitting the proper server URI.

Will file as a bug as I think it possibly is.

**Comment by nextgens** [comment]
The OOO responder is a dovecot/sieve functionality that runs on the "imap" container.

In ``/etc/dovecot/dovecot.conf``, it looks for ``submission_host`` and will connect to that. On master and 2.0 that should be ``front``, see https://github.com/Mailu/Mailu/blob/master/core/dovecot/conf/dovecot.conf#L8

``Connection timed out`` does mean that it was not reachable. I'd look into the network policies if I were you.

**Reply by jamesarbrown** [reply]
Hi + Thanks.

Its 2.0.30
It is connecting to front and front is the log i put up, as it is showing an SSL issue "empty". I presume connection timed out on the ```front``` is on the receiving end of the connection.

Two points why I dont think its a network policy

1/ I can connect to the smtp service as shown above, from within and outside of cluster (have network manager running in a pod in the namespace to test). The server is infact live and running and providing emails. Everything is running apart from vacation.

2/ I can get a valid certificate from the pods, which I feel is when nginx gets the correct URI imap.example.org, but not when its mailu-front.svc..... etc

So now I will go and look from the dovecot side of logging and see what it thinks is happening. 

I am stil thinking it is because nginx on the front does not have a certificate to give / wont recognise the URI dovecot is using.

**Reply by nextgens** [reply]
On 2.0 front runs nginx. The logs you have provided do not look like something nginx would log.

``Error: SSL context initialization failed, disabling SSL: Can't load SSL certificate (ssl_cert setting): The certificate is empty `` is not a bug, it's expected (that's dovecot saying it does not have a certificate).

Connect to the imap pod and try to reach front, see if that works.

**Reply by jamesarbrown** [reply]
no imap pod on 2.0.30

Its ghcr.io/mailu/dovecot:2.0.30 showing

Yes sorry..... looking at too many logs, it was infact dovecots logs all along

**Reply by jamesarbrown** [reply]
Later I will make a change to the DNS such the DNS goes direct to pod and ensure Traefik is not "fixing" stuff in the middle. Do not think it is as cant see how as its a HostSNI('*') and no TLS endpoint, but just need to check both of these repsonses are from exactly same pod.

```
mailu-dovecot-555d585554-ljzhk:/app# openssl s_client -connect mailu-front.mailu200124.svc.cluster.local:993
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 347 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

```

mailu-dovecot-555d585554-ljzhk:/app# openssl s_client -connect imap.example.org:993
CONNECTED(00000003)
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = example.org
verify return:1
---
Certificate chain
 0 s:CN = example.org
   i:C = US, O = Let's Encrypt, CN = R3
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jan 19 14:01:37 2024 GMT; NotAfter: Apr 18 14:01:36 2024 GMT
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Sep  4 00:00:00 2020 GMT; NotAfter: Sep 15 16:00:00 2025 GMT
 2 s:C = US, O = Internet Security Research Group, CN = ISRG Root X1
   i:O = Digital Signature Trust Co., CN = DST Root CA X3
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jan 20 19:14:03 2021 GMT; NotAfter: Sep 30 18:14:03 2024 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----

blah blah

---
read R BLOCK
* OK IMAP4 ready
```

**Reply by nextgens** [reply]
Okay, so try from the dovecot pod.

Forget about port 993, it's port 25 that matters.

``openssl s_client -connect mailu-front.mailu200124.svc.cluster.local:25 -starttls smtp``

**Reply by jamesarbrown** [reply]
Many thanks for the help

1/
```
mailu-dovecot-555d585554-ljzhk:/app# openssl s_client -connect mailu-front.mailu200124.svc.cluster.local:25 -starttls smtp
CONNECTED(00000003)
Didn't find STARTTLS in server response, trying anyway...
write:errno=32
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 33 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

2/
I have created a direct DNS entry mail.example.org and left the imap.example.org through traefik for now. Both hostnames are given to the helm chart anyway

Per https://github.com/Mailu/Mailu/discussions/3159#discussioncomment-8545637 now gives same response either way, zero ... so it does appear Traefik/indirect is/was covering up some other problem with SSL. Ie the "No peer certificate available response" as above


Will do some more checking this evening and try to get my head around it. I am still thinking there is a lack of an SSL cert response from the front pod, when I know they are there, mounted onto the pod in /certs

**Reply by jamesarbrown** [reply]
Think have found it. 

We have proxy protocol on for traefik. So if its not going through traefik, its not what ```front``` understands. So the fact postfix/dovecot/front are trying to go pod to pod directly is breaking this.

This is in the ```front``` section of helm

```
  ## @param front.extraEnvVars Extra environment variable to pass to the running container
  ## For example:
  ## extraEnvVars:
  ##  - name: MY_ENV_VAR
  ##    value: env_var_value
  extraEnvVars:
    - name: PROXY_PROTOCOL
      value: "mail"
    - name: REAL_IP_FROM
      value: "10.100.0.0/16,10.2.0.0/16"
```

**Reply by nextgens** [reply]
That's more like it.

Ok, we can create an issue at the back of it, ooo is probably broken when proxy protocol is in use for port 25.

**Reply by nextgens** [reply]
That's #3172

**Reply by jamesarbrown** [reply]
is there a way to work around / override submission in dovecot? I can see it in dovecot.conf so will see if I can test it, but obviously will reset if pod is killed.

**Reply by nextgens** [reply]
Yes, override ``submission_host`` with Traefik's FQDN and it should work

**Reply by jamesarbrown** [reply]
Tried and its complaining about DMARC, though all that is right and done in the domain... so will keep plugging away
```
vacation action: failed to send vacation response to james@example.org: <smtp(imap.example.org): DATA failed: 554 5.7.1 Rejected (anti-spoofing: DMARC compliance is enforced for local domains, regardless of the policy setting)> (permanent error)
```

**Reply by nextgens** [reply]
It sounds like the SUBNET definition is wrong. I am not sure what it's set to in helm charts.

**Reply by jamesarbrown** [reply]
Getting a bit past my understanding of the mechanics in the background.

The cluster pods are 10.42.0.0./16 and that is right. I have checked postconf on the postfix pod and that is right
```
mydomain = example.org
myhostname = example.org
mynetworks = 127.0.0.1/32 10.42.0.0/16
```

But obviously now dovecot is trying to get to this cluster via the normal network and through ```front``` so I think maybe just getting into a deeper and deeper mess. 

As a last try will add the actual network CIDR to postfix, but more suspect I am not understanding right

**Reply by nextgens** [reply]
Is your REAL_IP_FROM set to cover traefik's IP?

If not that is what needs fixing

**Reply by jamesarbrown** [reply]
Yes it was set. Treafik has two ip, frontend, backend. It was set as the backend ip 10.2.0.125 as mailu is on 10.2.0.90. 
Have tried frontend IP also, but no win

**Reply by jamesarbrown** [reply]
On further investigation Rspamd milter is doing this reject for postfix
```
DMARC_POLICY_REJECT (2) [example.org : No valid SPF, No valid DKIM,reject]
```
Yet the SPF is valid, and have checked via MXtoolbox
```
v=spf1 mx a:imap.example.org ip4:xxx.xxx.xxx.xxx/29 ~all
```

Think as postfix/rpsamd sees dovecots connection coming from an unintended path and is applying public rules for the milter and its tripping up the intention of working around the cluster ip subnet

**Reply by nextgens** [reply]
This means the rspamd/antispam pod does not see the DNS record.

**Reply by jamesarbrown** [reply]
This (Reject message) is only happening on the vacation response sent from Dovecot when there is the public URI override as submission_host, not happening for the incoming mails or within domain emails

So Antispam does have the visibility, just when its coming from Dovecot within the cluster via the public IP its failing spam checks.

**Healthy Check**
```
BAYES_HAM (-2.529402) [97.88%]
MIME_BASE64_TEXT_BOGUS (1)
MIME_HTML_ONLY (0.2)
MIME_BASE64_TEXT (0.1)
FROM_EQ_ENVFROM (0)
IS_LOCAL_DOMAIN_E (0) [example.org]
TO_DN_ALL (0)
MIME_TRACE (0) [0:~]
TO_MATCH_ENVRCPT_ALL (0)
ARC_NA (0)
RCVD_COUNT_ZERO (0) [0]
ASN (0) [asn:25310, ipnet:xx.xx.0.0/16, country:GB]
TO_DOM_EQ_FROM_DOM (0)
DKIM_SIGNED (0) [example.org:s=dkim]
FROM_HAS_DN (0)
NEURAL_HAM (0) [-0.895]
RCPT_COUNT_ONE (0) [1]
MID_RHS_MATCH_FROM (0)
IS_LOCAL_DOMAIN_H (0) [example.org]
```

**Reply by nextgens** [reply]
That sounds like a k8s/helm chart problem.

The load balancer should pass the source ip information: it's supposed to be an ip that's whitelisted (in ``SUBNET``: the imap/dovecot pod).

Which IP address does rspamd see? where is the information lost?

**Reply by jamesarbrown** [reply]
Hi,

Please enhance my understanding of the process, I saw the bug fix and I think I have tested that and proven it would not work, yet I do not think it a config issue on our part, but perhaps a concept issue.

1/ The IP Scheme would be
Our Loadbalancer (Proxy) > ```front``` (AKA Kubernetes Loadbalancer Proxy) > Cluster IP ```front``` service > postfix pod

2/ The Cluster IP subnet is the only one that can be trusted for EMAIL and it is in the helm chart. The ```front``` and therefore our LB can not be "email" trusted by default, as they are also passing public traffic.
```
## @param subnet Change this if you're using different address ranges for pods (IPv4)
subnet: 10.42.0.0/16
```
3/ Our Loadbalancer is trusted in a PROXY sense by this ?
```
  ## @param ingress.realIpFrom Sets the value of `REAL_IP_FROM` environment variable in the `front` pod
  realIpFrom: "10.2.0.125"
```

4/ ```front``` service is told to expect proxy (with last two lines commented out having read the documentation in more detail = "unset")
``` 
 ## @param front.extraEnvVars Extra environment variable to pass to the running container
  ## For example:
  ## extraEnvVars:
  ##  - name: MY_ENV_VAR
  ##    value: env_var_value
  extraEnvVars:
    - name: PROXY_PROTOCOL
      value: "mail"
    - name: TRAEFIK_VERSION
      value: "v2"
  #  - name: REAL_IP_FROM
  #    value: "10.2.0.0/16"
```

5/  I see the fix you have done, but is it not by nature of design, using the hostname it enters via the public email path, therefore it is subject to AUTH and RSPAMD via postfix master.cf etc?

6/ To follow on, if Dovecot is there trying to send an email to ourselves, calling itself "example.org", then that would indeed fail a DKIM check as its regarded as a spoof domain?

7/ My view was somehow to work around I would need to add another port on the postfix pod, creating a trusted incoming path and completely bypass the ```front``` by sending dovecot>postfix (which is not proxy at that level), rather than dovecot>lb>front>postfix as tried to do.

**Reply by nextgens** [reply]
There is a very simple concept in Mailu: "everything goes through front" ; while this wasn't completely true for 2.0 it is true for master.

I am unsure how the helm chart will be updated when the next release is out. All the various proxies should forward the source ip address... and that ip address will be authorized as it is the one of the pod hosting dovecot (that belongs to SUBNET). I have tested what's on master with docker and it works.

Regarding point 5, this is a feature and not a bug. Authorization is irrelevant (all of SUBNET is whitelisted) and we do want the anti-spam to run on what goes out via sieve.

@fastlorenzo may be able to shed some light on what will change in the chart.

**Reply by jamesarbrown** [reply]
Yes... i am watching the logs, the LB is passing the server IP, not the cluster IP. I suspect that is as many hops backwards as is achiveable. So then it falls outside of the whitelisted subnet. I will try some additional checks on this.

**Reply by jamesarbrown** [reply]
I just used from a networkmanager pod to ssh to each node as we have 2 nodes in each cluster.
If you ssh to same node that pod is running, sshd logs show the Cluster IP eg 10.42.0.0 subnet
If you ssh to another node, sshd logs show the IP of the node on which the pod is running in our case the 10.2.0.0/24 subnet  

It think this means that the ```front``` can only see the node IP in certain cases... so in kubernetes whitelisting and forcing to go via an external LB wont work?

![image](https://github.com/Mailu/Mailu/assets/891438/8191e10a-3d2c-4203-b8c2-3eb81611715f)

**Reply by jkolczasty** [reply]
Hello

I have similar problem. When proxy_protocol is used, front smtp/25 is expecting PROXY initialization. I works for load balancer, but when imap submission is trying to connect it fails with connect timeout, cause nginx is not sending banner, but is waiting for client initialization with proxy header.

Looks like using imap submission should use different port without proxy like webmail is using 10025.

**Reply by nextgens** [reply]
This is not an issue in master anymore; we are not planning to backport a fix since using PROXY on 2.0 is very difficult and will change significantly on master. Both with what's currently in and in case #3221 is merged

see
https://github.com/Mailu/Mailu/discussions/3159#discussioncomment-8544798

**Reply by jamesarbrown** [reply]
@nextgens We are waiting for Helm chart update on release to test. Have not looked recently but about 8 weeks ago was not changed from release we were running

**Reply by jkolczasty** [reply]
I'm using 2.0.41 and problem with submission from dovecot affects it as it uses 25 port with proxy_protocol enabled and times-out waiting for smtp banner and frontend nginx mail smtp port is waiting for proxy header - both endpoints are waiting for something and it timeouts.

**Reply by nextgens** [reply]
I have linked to the solution we use in master. In the meantime, define an override as follows:
```
submission_host = $fqdn_of_proxy
```
